chips/lowrisc/src/i2c.rs

```diff
@@ -0,0 +1,275 @@
+        regs.fifo_ctrl
+            .modify(FIFO_CTRL::RXILVL::RXLVL1 + FIFO_CTRL::FMTRST::SET);
+
+        // Enable all interrupts
+        regs.intr_enable.modify(
+            INTR::FMT_WATERMARK::SET
+                + INTR::RX_WATERMARK::SET
+                + INTR::FMT_OVERFLOW::SET
+                + INTR::RX_OVERFLOW::SET
+                + INTR::NAK::SET
+                + INTR::SCL_INTERFERENCE::SET
+                + INTR::SDA_INTERFERENCE::SET
+                + INTR::STRETCH_TIMEOUT::SET
+                + INTR::SDA_UNSTABLE::SET,
+        );
+
+        // Enable I2C Host
+        regs.ctrl.modify(CTRL::ENABLEHOST::SET);
+    }
+
+    fn disable(&self) {
+        let regs = self.registers;
+
+        regs.ctrl.modify(CTRL::ENABLEHOST::CLEAR);
+    }
+
+    fn write_read(&self, addr: u8, data: &'static mut [u8], write_len: u8, read_len: u8) {
+        let regs = self.registers;
+
+        // Zero out the LSB to signal a write
+        let write_addr = addr & !1;
+
+        // Set the start condition and the address
+        regs.fdata
+            .write(FDATA::START::SET + FDATA::FBYTE.val(write_addr as u32));
+
+        for i in 0..(write_len - 1) {
+            // Send the data
+            regs.fdata.write(FDATA::FBYTE.val(data[i as usize] as u32));
+        }
+
+        // Send the last byte with the stop signal
+        regs.fdata
+            .write(FDATA::FBYTE.val(data[write_len as usize] as u32) + FDATA::STOP::SET);
+
+        // Set the LSB to signal a read
+        let read_addr = addr | 1;
+
+        // Set the start condition and the address
+        regs.fdata
+            .write(FDATA::START::SET + FDATA::FBYTE.val(read_addr as u32));
+
+        for i in 0..read_len {
+            // Read the data
+            data[i as usize] = regs.rdata.read(RDATA::RDATA) as u8;
+        }
+    }
+
+    fn write(&self, addr: u8, data: &'static mut [u8], len: u8) {
+        let regs = self.registers;
+
+        // Zero out the LSB to signal a write
+        let write_addr = addr & !1;
+
+        // Set the start condition and the address
+        regs.fdata
+            .write(FDATA::START::SET + FDATA::FBYTE.val(write_addr as u32));
+
+        for i in 0..(len - 1) {
+            // Send the data
+            regs.fdata.write(FDATA::FBYTE.val(data[i as usize] as u32));
+        }
+
+        // Send the last byte with the stop signal
+        regs.fdata
+            .write(FDATA::FBYTE.val(data[len as usize] as u32) + FDATA::STOP::SET);
+    }
+
+    fn read(&self, addr: u8, buffer: &'static mut [u8], len: u8) {
+        let regs = self.registers;
+
+        // Set the LSB to signal a read
+        let read_addr = addr | 1;
+
+        // Set the start condition and the address
+        regs.fdata
+            .write(FDATA::START::SET + FDATA::FBYTE.val(read_addr as u32));
+
+        for i in 0..len {
+            // Read the data
+            buffer[i as usize] = regs.rdata.read(RDATA::RDATA) as u8;
+        }
```

Comment:
Will this actually work? I would think it would read much faster than bytes are received.